Formula/thrift.rb

```diff
@@ -59,6 +65,12 @@ def install
     ENV.deparallelize
     system "make"
     system "make", "install"
+
+    # Even when given a prefix to use it creates /usr/local/lib inside
+    # that dir & places the jars there, so we need to workaround that.
+    if build.with? "java"
+      libexec.install Dir.glob(buildpath/"usr/local/lib/*.jar")
```

Comment:
Is this a useful location for them?

Reply:
also `Dir["usr/local/lib/*.jar"]` should be good no?

Reply:
🤷‍♂️. Probably as useful as any. We don't permit `.jar` files to live nakedly in `lib` so they usually end up somewhere under `libexec`. Whether it's worth trying to write a wrapper script for any of them, I couldn't tell you, but we haven't to this point.

Reply:
> also Dir["usr/local/lib/*.jar"] should be good no?

I tried that and nothing got installed. Wasn't sure what I was screwing up 😅.

Reply:
Frankly that only one person has noticed/reported this option has been _extremely_ broken for a long time is probably telling in terms of interest. At least the `.jar`s are actually being installed now. I figure that's progress 😓.

Reply:
```
 1 | thrift                                                                                                                                                                                | 2,477 |  91.17%
 2 | thrift --with-java                                                                                                                                                                    |    42 |   1.55%
 3 | thrift --with-java --with-libevent --with-python                                                                                                                                      |    29 |   1.07%
 4 | thrift --HEAD --with-haskell --with-erlang --with-java --with-php --with-python                                                                                                       |    23 |   0.85%
 5 | thrift --HEAD --with-python --with-libevent                                                                                                                                           |    20 |   0.74%
 6 | thrift --with-php                                                                                                                                                                     |    18 |   0.66%
 7 | thrift --with-python                                                                                                                                                                  |    16 |   0.59%
 8 | thrift --with-java --with-python                                                                                                                                                      |    14 |   0.52%
 9 | thrift --HEAD                                                                                                                                                                         |     8 |   0.29%
10 | thrift --with-java --with-perl --with-python                                                                                                                                          |     8 |   0.29%
11 | thrift --with-java --with-php --with-python                                                                                                                                           |     8 |   0.29%
12 | thrift --with-haskell --with-erlang --with-java --with-php --with-libevent --with-python                                                                                              |     6 |   0.22%
13 | thrift --with-java --with-php --with-libevent --with-python                                                                                                                           |     6 |   0.22%
14 | thrift --with-haskell --with-erlang --with-java --with-perl --with-php --with-libevent --with-python                                                                                  |     5 |   0.18%
15 | thrift --with-erlang --with-java --with-php --with-libevent --with-python                                                                                                             |     4 |   0.15%
16 | thrift --with-haskell --with-erlang --with-perl --with-php --with-libevent --with-python                                                                                              |     4 |   0.15%
17 | thrift --with-java --with-libevent                                                                                                                                                    |     4 |   0.15%
18 | thrift --with-libevent                                                                                                                                                                |     4 |   0.15%
19 | thrift --with-erlang --with-perl --with-libevent --with-python                                                                                                                        |     3 |   0.11%
20 | thrift --with-libevent --with-python                                                                                                                                                  |     3 |   0.11%
21 | thrift --HEAD --with-java --with-perl --with-php --with-libevent --with-python                                                                                                        |     2 |   0.07%
22 | thrift --with-erlang                                                                                                                                                                  |     2 |   0.07%
23 | thrift --with-erlang --with-java --with-python                                                                                                                                        |     2 |   0.07%
24 | thrift --with-erlang --with-libevent                                                                                                                                                  |     2 |   0.07%
25 | thrift --with-haskell --with-java --with-libevent --with-python                                                                                                                       |     2 |   0.07%
26 | thrift --with-python --with-libevent                                                                                                                                                  |     2 |   0.07%
27 | thrift --with-haskell --with-erlang --with-java --with-perl --with-libevent --with-python                                                                                             |     1 |   0.04%
28 | thrift --with-haskell --with-erlang --with-php --with-libevent --with-python                                                                                                          |     1 |   0.04%
29 | thrift --with-php --with-python                                                                                                                                                       |     1 |   0.04%
============================================================================================================================================================================================================
Total                                                                                                                                                                                      | 2,717 |    100%
============================================================================================================================================================================================================
```
relatively popular :(

Reply:
How has nobody else come & pointed out how completely that option does not work currently. Yikes.

Reply:
They definitely get installed by
```
libexec.install Dir["usr/local/lib/*.jar"] if build.with? "java"
```

```
iMac-TMP:homebrew-core joe$ ls /usr/local/Cellar/thrift/0.11.0/libexec/
commons-codec-1.9.jar		junit-4.4.jar			mockito-all-1.9.5.jar
commons-logging-1.2.jar		libthrift-0.11.0-javadoc.jar	servlet-api-2.5.jar
httpclient-4.4.1.jar		libthrift-0.11.0.jar		slf4j-api-1.7.12.jar
httpcore-4.4.1.jar		log4j-1.2.17.jar		slf4j-log4j12-1.7.12.jar
iMac-TMP:homebrew-core joe$ 
```

Reply:
Strange. Wonder what I was typing locally earlier 😅. Pushed.

Reply:
>Probably as useful as any. 

An alternative might be `share/"java"` but I guess `libexec` works.

Reply:
Aye. `libexec` has become the default dumping ground for `jar` files over the years, but certainly can't remember any major debate as to how desirable that actually is.

Reply:
Dumping ground when we're env scripting though, right? I mean `gdal` even does `lib.install "gdal.jar"`

Reply:
> I mean gdal even does lib.install "gdal.jar"

Really? I thought we'd hard audited this out of existence. Although the audit needs running after installation, so if it's not a default option it'll have never been spotted.

Reply:
[Definitely not supposed to be there though](https://github.com/Homebrew/brew/blob/76cd7c79498daf380d99ed7379394618164b602b/Library/Homebrew/formula_cellar_checks.rb#L43-L57).

Reply:
Yeah that one is still probably an audit violation along with all the hard-coded python framework paths that don't get fixed because they're non-default :)

Reply:
Yeah, plenty of those around, that's for sure. I might look at `gdal` later because that is a bit naughty & definitely not an example others need to stumble upon and think "ooh handy".

Reply:
OK, so `collectd --with-java` gets you
```
iMac-TMP:~ joe$ find /usr/local/Cellar/collectd/5.8.0|grep jar
/usr/local/Cellar/collectd/5.8.0/share/collectd/java/collectd-api.jar
/usr/local/Cellar/collectd/5.8.0/share/collectd/java/generic-jmx.jar
```

Reply:
`subversion --with java` gets you
```
iMac-TMP:~ joe$ find /usr/local/Cellar/subversion/1.9.7_2|grep jar
/usr/local/Cellar/subversion/1.9.7_2/lib/svn-javahl/svn-javahl.jar
```
I'm really enjoying the consistency here.

Reply:
`open-babel --with-java` plops it in `lib`
```
iMac-TMP:~ joe$ find /usr/local/Cellar/open-babel/2.4.1|grep jar
/usr/local/Cellar/open-babel/2.4.1/lib/openbabel.jar
iMac-TMP:~ joe$ 
```

Reply:
`portmidi --with-java` and `opencolorio --with-java` both die in balls of flames

Reply:
`ibex --with java` gets you

```
iMac-TMP:~ joe$ find /usr/local/Cellar/ibex/2.6.5|grep jar
/usr/local/Cellar/ibex/2.6.5/share/java/ibex.jar
```

Reply:
`erlang --with java` has its own ideas
```
iMac-TMP:~ joe$ find /usr/local/Cellar/erlang/20.2.2|grep jar
/usr/local/Cellar/erlang/20.2.2/lib/erlang/lib/ic-4.4.3/priv/ic.jar
/usr/local/Cellar/erlang/20.2.2/lib/erlang/lib/jinterface-1.8.1/priv/OtpErlang.jar
```

Reply:
and `ice --with-java`
```
iMac-TMP:~ joe$ find /usr/local/Cellar/ice/3.7.0_2|grep jar
/usr/local/Cellar/ice/3.7.0_2/IceGrid GUI.app/Contents/Java/icegridgui.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/glacier2-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/glacier2-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/glacier2-compat-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/glacier2-compat-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/ice-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/ice-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/ice-compat-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/ice-compat-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icebox-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icebox-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icebox-compat-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icebox-compat-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icebt-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icebt-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icebt-compat-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icebt-compat-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icediscovery-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icediscovery-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icediscovery-compat-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icediscovery-compat-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icegrid-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icegrid-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icegrid-compat-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icegrid-compat-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icegridgui.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icelocatordiscovery-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icelocatordiscovery-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icelocatordiscovery-compat-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icelocatordiscovery-compat-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icepatch2-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icepatch2-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icepatch2-compat-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icepatch2-compat-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icessl-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icessl-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icestorm-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icestorm-3.7.0.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icestorm-compat-3.7.0-sources.jar
/usr/local/Cellar/ice/3.7.0_2/share/java/icestorm-compat-3.7.0.jar
```

Reply:
`share/java` seems to be common. We could adhere to that for libraries, but I'm concerned about the audit message's intention and doing that, because the two seem to contradict each other. The audit failure essentially is saying don't place Java libraries anywhere public because they often have common names & that can cause issues, and `share/java` is a public prefix.

I think my order of preference would be `libexec`, `pkgshare/java` and `lib/name`.

Reply:
Of everything naturally happening I feel like `collectd` is potentially screwing us over the least.

Reply:
It seems weird to put things in libexec that are intended for users to do something with directly, given that libexec is supposed to be private and internal to the formula, and doesn't get symlinked at all. The weird gnubin directories are a notable exception, but the caveats tell people they're there and what to do with them.

Reply:
https://github.com/Homebrew/brew/blob/master/docs/Formula-Cookbook.md#variables-for-directory-locations

>Note that in the context of Homebrew, libexec is reserved for private use by the formula and therefore is not symlinked into HOMEBREW_PREFIX.

Putting Java bindings people are supposed to use in there seems to violate that description, no?

Reply:
>I think my order of preference would be libexec, pkgshare/java and lib/name.

Yet another option would be a `java_share` equal to `#{share}/java/#{name}`

Reply:
>  given that libexec is supposed to be private and internal to the formula

I agree that `libexec` isn't perfect either. It is the audit's advice though, even if the audit makes a vague hint at that mostly being intended for stuff that's going to get wrapped into `bin`, so maybe we need to tweak the audit wording if we change things broadly. Not saying that isn't a worthwhile change, just something I don't want to forget.

> Yet another option would be a java_share equal to #{share}/java/#{name}

I'd... be okay with this, if you have issues with `pkgshare`.

Reply:
`pkgshare/java` seems reasonable to me

Reply:
I moved them both there, and `gdal` in the other PR. We can try to standardised Java libraries there from now on.